### ISSUE TITLE
Fix and/or document watch window constraint

### ISSUE THREAD

**Issue by ssergiienko** [kubernetes/kubernetes#16429]
Kubernetes 1.0.3
Watch api is not clear and almost useless. Also it does not comply documentation. Will try to explain by examples:

```
curl localhost:8080/api/v1/pods?watch=true
```

Returns only one latest object for each pod with:
...
"resourceVersion":"41195"
...
"resourceVersion":"43383"
...
Then:

```
# with any resourceVersion lower the last at least by 1 or more returns
curl localhost:8080/api/v1/pods?watch=true\&resourceVersion=43300
```

``` json
{"type":"ERROR","object":{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"401: The event in requested index is outdated and cleared (the requested history has been cleared [95462/43300]) [95461]"}}
```

At the same time for nodes(which is updated every few seconds) same approach is works:

```
curl localhost:8080/api/v1/nodes?watch=true
# few objects with
"resourceVersion":"95466"
```

```
curl localhost:8080/api/v1/nodes?watch=true\&resourceVersion=95000
# returns 466 object entries as expected
```

Why whis happens? resourveVersion is a global counter for all `kind`'s in kubernetes? If yes, where this is described, and how this api can be useful? I read https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#concurrency-control-and-consistency but it's said nothing about this(or i missed this).
So how i can retrieve any changes for pods if counter always in future (for pods or any infrequently changed objects)? Node's changes will almost always "shadow" any others.

According to the swagger docs (http://kubernetes.io/third_party/swagger-ui/#!/api%2Fv1/watchServiceList):

```
resourceVersion:

When specified with a watch call, shows changes that occur after that particular version of a resource. Defaults to changes from the beginning of history.
```

But this is also not true! When resourceVersion is omitted it returns only last changes, not `from the beginning of history` as described in docs, and this even confirmed in https://github.com/kubernetes/kubernetes/issues/13969

**Comment by ssergiienko** [comment]
This propose https://github.com/kubernetes/kubernetes/issues/7018#issuecomment-101079247 is also suitable in our case, hope it will be implemented even in api v1 (adding new field breaks nothing)

**Comment by lavalamp** [comment]
You should list first to get the complete state and a resource version. Then you watch with that resource version, and you are guaranteed not to miss changes.

>  resourveVersion is a global counter for all kind's in kubernetes? 

This happens to be no longer true, it is an implementation detail that you should not rely on.

There's no guarantees about watching without first listing. It's useful only in certain circumstances.

**Comment by ssergiienko** [comment]
@lavalamp
> You should list first to get the complete state and a resource version. Then you watch with that resource version, and you are guaranteed not to miss changes.

I list objects with `curl localhost:8080/api/v1/pods?watch=true` and see that `"resourceVersion":"43383"` is greater than my last processed... for example 43000... Then i call: 
```
curl localhost:8080/api/v1/pods?watch=true\&resourceVersion=43001
```
and instead of objects listing starting from 43300 i get:
```
{"type":"ERROR","object":{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"401: The event in requested index is outdated and cleared (the requested history has been cleared [95462/43301]) [95461]"}}
```
What am i doing wrong? Or what do you mean by `You should list first to get the complete state and a resource version`  ?

If i try to start listing pods from very latest resourceVersion of cluster (retrieved from api/nodes) than request is just blocked waiting:
```
[root@static ~]# curl localhost:8080/api/v1/pods?watch=true\&resourceVersion=95410
^C
```

**Comment by lavalamp** [comment]
First `curl localhost:8080/api/v1/pods`. Find the ResourceVersion in the
list meta.

Then `curl localhost:8080/api/v1/pods?watch=true&resourceVersion=xxxx`.

**Comment by ssergiienko** [comment]
Yap, that was it! Thank you very match! I thought that i must use resourceVersion of objects but it turns out that i must use resourceVersion of list itself instead... Maybe some clarification/example in docs about this things is needed?

Please, correct me: For now(v1.0.3), if i have more than 1000 change events for `nodes` 
since my last `GET /pods` than i CAN'T see all changes for pods after reconnect because they are already lost (too old) ? If this `resourceVersion` counter is global for all kinds (for now) than it will turn out that way. Or i am wrong?

Anyway, thanks for help!

**Comment by lavalamp** [comment]
Yes, that's correct. You can't take too long between the LIST and WATCH calls. I think in the future, the watch window will be 1000 changes per resource.

**Comment by lavalamp** [comment]
This is actually documented:

> "Watch" operations specify resourceVersion using a query parameter. It is used to specify the point at which to begin watching the specified resources. This may be used to ensure that no mutations are missed between a GET of a resource (or list of resources) and a subsequent Watch, even if the current version of the resource is more recent. This is currently the main reason that list operations (GET on a collection) return resourceVersion.

https://github.com/kubernetes/kubernetes/blob/c28ddf9ff8632bf5d545a0de7d84509214f45ea0/docs/devel/api-conventions.md#L469

We're willing to take PRs to improve this, though.

**Comment by evanphx** [comment]
I'm seeing the same behavior using the API as documented. Namely fetching `/api/v1/namespaces/endpoints/metrics`, the object returned has `"ResourceVersion": "1694561"`. If I turn around and issue a watch with `/api/v1/watch/namespaces/endpoints/metrics?resourceVersion= 1694561` I get the error:

```
{"kind":"Status","apiVersion":"v1","metadata":{"selfLink":"/api/v1/namespaces/default/endpoints/metrics"},"status":"Failure","message":"401: The event in requested index is outdated and cleared (the requested history has been cleared [1753236/1694562]) [1754235]"}
```

I presume the issue is that this particular endpoint has not been changed in a while (thats true) and the revision it was changed has been compacted by etcd is no longer available for watching. Kubernetes should be abstracting this error by updating the objects revisionVersion to the current etcd revision wavefront so that they're still watchable.

Basically, kubernetes embeds etcd revision numbers in objects but etcd invalidates the usage of those stored revision numbers, breaking kubernetes APIs.

**Comment by evanphx** [comment]
One more quick comment: This isn't a documentation issue. It's a bug in the interaction between the client, apiServer, and etcd.

**Comment by bgrant0607** [comment]
Removed documentation label. Documentation issue is #15666

**Comment by smarterclayton** [comment]
Closing due to age, reopen if the doc issue doesn't resolve.
